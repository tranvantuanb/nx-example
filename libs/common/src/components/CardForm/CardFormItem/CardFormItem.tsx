import React, { useContext } from 'react';
import tw, { styled } from 'twin.macro';
import { Row, Col, ColProps, Form, FormItemProps } from 'antd';

import { CardFormContext } from '../CardFormContext';

const StyledFormItem = styled(Form.Item)`
  ${tw`mb-0`}
`;

export interface CardFormItemProps extends FormItemProps {
  /**
   * This function is to use for rendering custom UI in view mode.
   * E.g:
   * - render the form value into different formats
   * - render File Viewer
   */
  renderViewValue?: (value) => React.ReactNode;
  /**
   * This prop allows customization of the FormItem Col.
   * E.g: if we want to render 1 field as full width, pass in
   * colProps = { xs: 24 }
   */
  colProps?: ColProps;
}

const FormItemViewOnly: React.FC<CardFormItemProps> = React.memo(
  ({ renderViewValue = (val) => val, label, name, initialValue }) => {
    const form = Form.useFormInstance();
    const fieldValue = (name ? form.getFieldValue(name) : '') || initialValue;
    const viewValue = fieldValue ? renderViewValue(fieldValue) : '';

    return (
      <Row className="card-form-item-view-only flex">
        <Col span={8} className="text-sm font-bold text-black-300">
          {label}
        </Col>
        <Col span={16} className="text-sm font-bold truncate overflow-ellipsis">
          {viewValue}
        </Col>
      </Row>
    );
  }
);

export const CardFormItem: React.FC<CardFormItemProps> = React.memo(
  ({ className, colProps = {}, children, renderViewValue, ...otherProps }) => {
    const { isEditing } = useContext(CardFormContext);

    const mergedColProps = {
      span: isEditing ? 12 : 24,
      ...colProps,
    };

    return (
      <Col className={className} {...mergedColProps}>
        {!isEditing ? (
          <FormItemViewOnly renderViewValue={renderViewValue} {...otherProps} />
        ) : (
          <StyledFormItem {...otherProps}>{children}</StyledFormItem>
        )}
      </Col>
    );
  }
);
