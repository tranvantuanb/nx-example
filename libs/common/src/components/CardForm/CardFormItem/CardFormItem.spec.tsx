import { render } from '../../../test';

import { CardFormItem, CardFormItemProps } from './CardFormItem';

describe('CardFormItem', () => {
  it('should render successfully', () => {
    const defaultProps: CardFormItemProps = {};
    const { container } = render(<CardFormItem {...defaultProps} />);
    expect(container).toBeTruthy();
    expect(container).toMatchSnapshot();
  });
});
