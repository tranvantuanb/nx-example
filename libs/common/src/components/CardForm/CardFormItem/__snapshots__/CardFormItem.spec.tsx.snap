// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardFormItem should render successfully 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  font-size: 14px;
  line-height: 14px;
  font-weight: 700;
  --tw-text-opacity: 1;
  color: rgba(161,167,188,var(--tw-text-opacity));
}

.c2 {
  font-size: 14px;
  line-height: 14px;
  font-weight: 700;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
}

<div>
  <div
    class="ant-col ant-col-24"
  >
    <div
      class="ant-row c0 card-form-item-view-only"
    >
      <div
        class="ant-col ant-col-8 c1"
      />
      <div
        class="ant-col ant-col-16 c2"
      />
    </div>
  </div>
</div>
`;
